src/core/mempool.rs

```diff
@@ -535,6 +581,7 @@ const MEMPOOL_INDEXES: &'static [&'static str] = &[
 
 pub struct MemPoolDB {
     pub db: DBConn,
+    db2: Arc<Mutex<DBConn>>,
```

Comment:
note: added Arc<Mutex> to extra DbConn.. I actually think this makes sense, because it is like a tiny (size 1) db connection pool.

i had to do this to pass a DBConn into the "transactions in the mempool" iterator, without borrowing the reference to `self=MemPoolDB`. this is because the class also has to be borrowed inside the loop as mutable.

rust doesn't let you borrow as both mutable and immutable at the same time.

Reply:
The Rust-y thing to do here would be to make `db` into `Option<DBConn>`, and then add a method like this:

```rust
fn with_db<F, R>(&mut self, func: F) -> R
where
   F: FnOnce(&mut MemPoolDB, &mut DBConn) -> R
{
   let db = self.db.take().expect("BUG: db is already taken");
   let res = func(self, db);
   self.db = Some(db);
   res
}
```

Then, you put code that needs mutable references to `self.db` and `self` into `func` like so:

```rust
let result = self.with_db(|mempool, db| {
   // this compiles since mempool is &mut MemPollDB
   mempool.path = "foo".to_string();
   // this also compiles since db is &mut DBConn
   let mut tx = tx_begin_immediate(db)?;
   /* ... etc. ... */
});
```

We do this with `PeerNetwork` in a lot of places, for example, in order to get mutable access to both the `PeerNetwork` and one of its inner state machines.

`Arc<..>` is typically only used for sharing state between threads, which is not happening here.

Reply:
If we want to say that MemPoolDB doesn't spawn, that would simplify things.

I think you solution would be incorrect if there was a spawned thread, but not mutex.

Reply:
It sounds like you need to re-think the factoring of your test.  Putting a second database connection into the mempool DB struct just to get around a compile error in your PR indicates that you're probably doing something wrong.

Reply:
actually.. I added a new TestPeer test in some branch, but it's not in this branch.

Thus, we can definitely say `TestPeer` is spawning already, and if that's a problem, we should unwind it.